diff -ruNa linux-3.2.0-psp04.06.00.11tar/.config linux-3.2.0-psp04.06.00.11/.config
--- linux-3.2.0-psp04.06.00.11tar/.config	2017-09-13 01:10:51.000000000 -0700
+++ linux-3.2.0-psp04.06.00.11/.config	2020-09-04 06:01:42.036143206 -0700
@@ -570,7 +570,8 @@
 # CONFIG_NF_CONNTRACK_FTP is not set
 # CONFIG_NF_CONNTRACK_H323 is not set
 # CONFIG_NF_CONNTRACK_IRC is not set
-# CONFIG_NF_CONNTRACK_NETBIOS_NS is not set
+CONFIG_NF_CONNTRACK_BROADCAST=y
+CONFIG_NF_CONNTRACK_NETBIOS_NS=y
 # CONFIG_NF_CONNTRACK_SNMP is not set
 # CONFIG_NF_CONNTRACK_PPTP is not set
 # CONFIG_NF_CONNTRACK_SANE is not set
@@ -658,7 +659,7 @@
 CONFIG_NF_NAT_NEEDED=y
 CONFIG_IP_NF_TARGET_MASQUERADE=y
 # CONFIG_IP_NF_TARGET_NETMAP is not set
-# CONFIG_IP_NF_TARGET_REDIRECT is not set
+CONFIG_IP_NF_TARGET_REDIRECT=y
 # CONFIG_NF_NAT_FTP is not set
 # CONFIG_NF_NAT_IRC is not set
 # CONFIG_NF_NAT_TFTP is not set
@@ -1137,7 +1138,7 @@
 CONFIG_USB_NET_NET1080=m
 # CONFIG_USB_NET_PLUSB is not set
 # CONFIG_USB_NET_MCS7830 is not set
-# CONFIG_USB_NET_RNDIS_HOST is not set
+CONFIG_USB_NET_RNDIS_HOST=y
 CONFIG_USB_NET_CDC_SUBSET=m
 # CONFIG_USB_ALI_M5632 is not set
 # CONFIG_USB_AN2720 is not set
diff -ruNa linux-3.2.0-psp04.06.00.11tar/drivers/usb/class/cdc-acm.c linux-3.2.0-psp04.06.00.11/drivers/usb/class/cdc-acm.c
--- linux-3.2.0-psp04.06.00.11tar/drivers/usb/class/cdc-acm.c	2013-06-25 14:38:13.000000000 -0700
+++ linux-3.2.0-psp04.06.00.11/drivers/usb/class/cdc-acm.c	2020-09-04 07:09:26.732287694 -0700
@@ -1162,6 +1162,10 @@
 				usb_sndbulkpipe(usb_dev, epwrite->bEndpointAddress),
 				NULL, acm->writesize, acm_write_bulk, snd);
 		snd->urb->transfer_flags |= URB_NO_TRANSFER_DMA_MAP;
+		#if 1 //Added by Quectel for zero packet
+		if (usb_dev->descriptor.idVendor == 0x1519 && usb_dev->descriptor.idProduct == 0x0020)
+			snd->urb->transfer_flags |= URB_ZERO_PACKET; 		
+		#endif
 		snd->instance = acm;
 	}
 
@@ -1220,6 +1224,14 @@
 
 	acm_table[minor] = acm;
 
+	#if 1 //Added by Quectel //For USB Auto Suspend & USB Remote Wakeup
+	if((usb_dev->descriptor.idVendor == 0x1519) && (usb_dev->descriptor.idProduct == 0x0020)) {
+		//pm_runtime_set_autosuspend_delay(&usb_dev->dev, 3000);
+		//usb_enable_autosuspend(usb_dev); 
+		//device_init_wakeup(&usb_dev->dev, 1); //usb remote wakeup
+	}
+	#endif
+
 	return 0;
 alloc_fail7:
 	for (i = 0; i < ACM_NW; i++)
diff -ruNa linux-3.2.0-psp04.06.00.11tar/drivers/usb/serial/option.c linux-3.2.0-psp04.06.00.11/drivers/usb/serial/option.c
--- linux-3.2.0-psp04.06.00.11tar/drivers/usb/serial/option.c	2017-07-19 20:58:30.000000000 -0700
+++ linux-3.2.0-psp04.06.00.11/drivers/usb/serial/option.c	2020-09-04 07:01:42.660271198 -0700
@@ -307,15 +307,18 @@
 #define BANDRICH_PRODUCT_1012			0x1012
 
 #define QUALCOMM_VENDOR_ID			0x05C6
-
-
-//add by alan 150820
 #define QUECTEL_VENDOR_ID 			0x2C7C
 #define QUECTEL_PRODUCT_UC15 		0x9090
 #define QUECTEL_PRODUCT_UC20 		0x9003
-#define QUECTEL_PRODUCT_EC20 		0x9215
 #define QUECTEL_PRODUCT_EC25 		0x0125
 #define QUECTEL_PRODUCT_EC21 		0x0121
+#define QUECTEL_PRODUCT_EC20 		0x9215
+#define QUECTEL_PRODUCT_EG91 		0x0191
+#define QUECTEL_PRODUCT_EG95 		0x0195
+#define QUECTEL_PRODUCT_EG06_EP06_EM06 	0x0306
+#define QUECTEL_PRODUCT_BG96 		0x0296
+#define QUECTEL_PRODUCT_AG35 		0x0435
+#define QUECTEL_PRODUCT_EC200S 		0x6002
 
 #define CMOTECH_VENDOR_ID			0x16d8
 #define CMOTECH_PRODUCT_6008			0x6008
@@ -776,17 +779,19 @@
 	{ USB_DEVICE(QUALCOMM_VENDOR_ID, 0x6613)}, /* Onda H600/ZTE MF330 */
 	{ USB_DEVICE(QUALCOMM_VENDOR_ID, 0x9000)}, /* SIMCom SIM5218 */
 
-	//add by alan 150820
-	{ USB_DEVICE(QUALCOMM_VENDOR_ID, QUECTEL_PRODUCT_UC15)}, /* QUECTEL
-	UC15 */
-	{ USB_DEVICE(QUALCOMM_VENDOR_ID, QUECTEL_PRODUCT_UC20)}, /* QUECTEL
-	UC20 */
-	{ USB_DEVICE(QUALCOMM_VENDOR_ID, QUECTEL_PRODUCT_EC20)}, /* QUECTEL
-	EC20 */
-	{ USB_DEVICE(QUECTEL_VENDOR_ID, QUECTEL_PRODUCT_EC25)}, /* QUECTEL EC25
-	*/
-	{ USB_DEVICE(QUECTEL_VENDOR_ID, QUECTEL_PRODUCT_EC21)}, /* QUECTEL EC21
-	*/
+	#if 1 //Added by Quectel
+	{ USB_DEVICE(QUALCOMM_VENDOR_ID, QUECTEL_PRODUCT_UC15) }, /* Quectel UC15 */ 
+	{ USB_DEVICE(QUALCOMM_VENDOR_ID, QUECTEL_PRODUCT_UC20) }, /* Quectel UC20 */
+	{ USB_DEVICE(QUECTEL_VENDOR_ID,  QUECTEL_PRODUCT_EC25) }, /* Quectel EC25 */
+	{ USB_DEVICE(QUECTEL_VENDOR_ID, QUECTEL_PRODUCT_EC21) }, /* Quectel EC21 */
+	{ USB_DEVICE(QUALCOMM_VENDOR_ID, QUECTEL_PRODUCT_EC20) }, /* Quectel EC20 */
+	{ USB_DEVICE(QUECTEL_VENDOR_ID, QUECTEL_PRODUCT_EG91) }, /* Quectel EG91 */
+	{ USB_DEVICE(QUECTEL_VENDOR_ID, QUECTEL_PRODUCT_EG95) }, /* Quectel EG95 */
+	{ USB_DEVICE(QUECTEL_VENDOR_ID, QUECTEL_PRODUCT_EG06_EP06_EM06) }, /* Quectel EG06/EP06/EM06 */ 
+	{ USB_DEVICE(QUECTEL_VENDOR_ID, QUECTEL_PRODUCT_BG96) }, /* Quectel BG96 */
+	{ USB_DEVICE(QUECTEL_VENDOR_ID, QUECTEL_PRODUCT_AG35) }, /* Quectel AG35 */ 
+	{ USB_DEVICE(QUECTEL_VENDOR_ID, QUECTEL_PRODUCT_EC200S) }, /* Quectel EC200S */ 
+	#endif
 
 
 	{ USB_DEVICE(CMOTECH_VENDOR_ID, CMOTECH_PRODUCT_6280) }, /* BP3-USB & BP3-EXT HSDPA */
@@ -1269,9 +1274,9 @@
 /* per port private data */
 
 #define N_IN_URB 4
-#define N_OUT_URB 4
+#define N_OUT_URB 4 //Quectel 1
 #define IN_BUFLEN 4096
-#define OUT_BUFLEN 4096
+#define OUT_BUFLEN 4096 //Quectel 128
 
 struct option_port_private {
 	/* Input endpoints and buffer for this port */
@@ -1383,26 +1388,48 @@
 		return -ENODEV;
 
 
-	// Add to cancel the binding of UC20 NDIS interface -> start
-	if (serial->dev->descriptor.idVendor == 0x05c6 &&
-	serial->dev->descriptor.idProduct == 0x9003 &&
-	serial->interface->cur_altsetting->desc.bInterfaceNumber == 4)
-	return -ENODEV;
-	// Add to cancel the binding of UC20 NDIS interface -> end
-	// Add to cancel the binding of EC20 NDIS interface -> start
-	if (serial->dev->descriptor.idVendor == 0x05c6 &&
-	serial->dev->descriptor.idProduct == 0x9215 &&
-	serial->interface->cur_altsetting->desc.bInterfaceNumber == 4)
-	return -ENODEV;
-	// Add to cancel the binding of EC20 NDIS interface -> end
-	// Add to cancel the binding of EC25 NDIS interface -> start
-	if (serial->dev->descriptor.idVendor == 0x2C7C &&
-	serial->interface->cur_altsetting->desc.bInterfaceNumber == 4)
-	return -ENODEV;
-	// Add to cancel the binding of EC25 NDIS interface -> end
-
-
+	#if 1 //Added by Quectel
 
+	if (serial->dev->descriptor.idVendor == cpu_to_le16(0x05c6)) {
+		if (serial->dev->descriptor.idProduct == cpu_to_le16(0x9090)) {	//UC15
+			//For USB Auto Suspend & USB Remote Wakeup
+			//pm_runtime_set_autosuspend_delay(&serial->dev->dev, 3000); 
+			//usb_enable_autosuspend(serial->dev);
+			//device_init_wakeup(&serial->dev->dev, 1); //usb remote wakeup
+		} else if (serial->dev->descriptor.idProduct == cpu_to_le16(0x9003)) {	//UC20
+			if (serial->interface->cur_altsetting->desc.bInterfaceNumber >= 4) {
+				return -ENODEV;
+			}
+			//For USB Auto Suspend & USB Remote Wakeup
+			//pm_runtime_set_autosuspend_delay(&serial->dev->dev, 3000);
+			//usb_enable_autosuspend(serial->dev); 
+			//device_init_wakeup(&serial->dev->dev, 1); //usb remote wakeup
+		} else if (serial->dev->descriptor.idProduct == cpu_to_le16(0x9215)) {	//EC20
+			if (serial->interface->cur_altsetting->desc.bInterfaceNumber >= 4) {
+				return -ENODEV;
+			}
+			//For USB Auto Suspend & USB Remote Wakeup
+			//pm_runtime_set_autosuspend_delay(&serial->dev->dev, 3000);
+			//usb_enable_autosuspend(serial->dev); 
+			//device_init_wakeup(&serial->dev->dev, 1); //usb remote wakeup
+		}
+	} else if (serial->dev->descriptor.idVendor == cpu_to_le16(0x2C7C)) {
+		if (serial->dev->descriptor.idProduct == cpu_to_le16(0x6002)) {	//EC200S
+			if (serial->interface->cur_altsetting->desc.bInterfaceClass != 0xFF) {
+				return -ENODEV;
+			}
+		} else {	//EC25 & EC21 & EG91 & EG95 & EG06_EP06_EM06 BG96 AG35
+			if (serial->interface->cur_altsetting->desc.bInterfaceNumber >= 4) {
+				return -ENODEV;
+			}
+			//For USB Auto Suspend & USB Remote Wakeup
+			//pm_runtime_set_autosuspend_delay(&serial->dev->dev, 3000); 
+			//usb_enable_autosuspend(serial->dev);
+			//device_init_wakeup(&serial->dev->dev, 1); //usb remote wakeup
+		}
+	}
+	
+	#endif
 
 	data = serial->private = kzalloc(sizeof(struct usb_wwan_intf_private), GFP_KERNEL);
 	if (!data)
diff -ruNa linux-3.2.0-psp04.06.00.11tar/drivers/usb/serial/qcserial.c linux-3.2.0-psp04.06.00.11/drivers/usb/serial/qcserial.c
--- linux-3.2.0-psp04.06.00.11tar/drivers/usb/serial/qcserial.c	2013-06-25 14:38:13.000000000 -0700
+++ linux-3.2.0-psp04.06.00.11/drivers/usb/serial/qcserial.c	2020-09-04 06:01:42.036143206 -0700
@@ -63,7 +63,7 @@
 	{USB_DEVICE(0x03f0, 0x241d)},	/* HP Gobi 2000 QDL device (VP412) */
 	{USB_DEVICE(0x03f0, 0x251d)},	/* HP Gobi 2000 Modem device (VP412) */
 	{USB_DEVICE(0x05c6, 0x9214)},	/* Acer Gobi 2000 QDL device (VP413) */
-	{USB_DEVICE(0x05c6, 0x9215)},	/* Acer Gobi 2000 Modem device (VP413) */
+	//{USB_DEVICE(0x05c6, 0x9215)},	/* Acer Gobi 2000 Modem device (VP413) */
 	{USB_DEVICE(0x05c6, 0x9264)},	/* Asus Gobi 2000 QDL device (VR305) */
 	{USB_DEVICE(0x05c6, 0x9265)},	/* Asus Gobi 2000 Modem device (VR305) */
 	{USB_DEVICE(0x05c6, 0x9234)},	/* Top Global Gobi 2000 QDL device (VR306) */
diff -ruNa linux-3.2.0-psp04.06.00.11tar/drivers/usb/serial/usb-serial.c linux-3.2.0-psp04.06.00.11/drivers/usb/serial/usb-serial.c
--- linux-3.2.0-psp04.06.00.11tar/drivers/usb/serial/usb-serial.c	2016-08-23 00:08:52.000000000 -0700
+++ linux-3.2.0-psp04.06.00.11/drivers/usb/serial/usb-serial.c	2020-09-04 06:01:42.040143206 -0700
@@ -50,6 +50,9 @@
 	.disconnect =	usb_serial_disconnect,
 	.suspend =	usb_serial_suspend,
 	.resume =	usb_serial_resume,
+	#if 1 //Added by Quectel
+	.reset_resume = usb_serial_resume,
+	#endif
 	.no_dynamic_id = 	1,
 	.supports_autosuspend =	1,
 };
diff -ruNa linux-3.2.0-psp04.06.00.11tar/drivers/usb/serial/usb_wwan.c linux-3.2.0-psp04.06.00.11/drivers/usb/serial/usb_wwan.c
--- linux-3.2.0-psp04.06.00.11tar/drivers/usb/serial/usb_wwan.c	2013-06-25 14:38:13.000000000 -0700
+++ linux-3.2.0-psp04.06.00.11/drivers/usb/serial/usb_wwan.c	2020-09-04 06:22:03.808186636 -0700
@@ -482,6 +482,21 @@
 			  usb_sndbulkpipe(serial->dev, endpoint) | dir,
 			  buf, len, callback, ctx);
 
+	#if 1 //Added by Quectel for zero packet 
+	if (dir == USB_DIR_OUT) {
+		struct usb_device_descriptor *desc = &serial->dev->descriptor;
+		if (desc->idVendor == cpu_to_le16(0x05C6)) {
+			if (desc->idProduct == cpu_to_le16(0x9090) || 
+			desc->idProduct == cpu_to_le16(0x9003) || 
+			desc->idProduct == cpu_to_le16(0x9215)) {
+				urb->transfer_flags |= URB_ZERO_PACKET;
+			}
+		} else if  (desc->idVendor == cpu_to_le16(0x2C7C)) {
+			urb->transfer_flags |= URB_ZERO_PACKET;
+		}
+	}
+	#endif
+
 	return urb;
 }
 
