--- buildroot-2019.02.11.orgi/package/python-ipy/Config.in	2021-02-23 21:25:48.352711000 +0800
+++ buildroot-2019.02.11/package/python-ipy/Config.in	2021-02-23 21:21:38.535317000 +0800
@@ -1,6 +1,6 @@
 config BR2_PACKAGE_PYTHON_IPY
 	bool "python-ipy"
-	depends on BR2_PACKAGE_PYTHON
+	#depends on BR2_PACKAGE_PYTHON
 	help
 	  IPy - class and tools for handling of IPv4 and
 	  IPv6 addresses and networks.
